cmd/queue/main.go

```diff
@@ -260,53 +260,50 @@ func main() {
 	probe := buildProbe(logger, env.ServingReadinessProbe)
 	healthState := &health.State{}
 
-	server := buildServer(ctx, env, healthState, probe, stats, logger)
-	adminServer := buildAdminServer(logger, healthState)
-	metricsServer := buildMetricsServer(promStatReporter, protoStatReporter)
-
-	servers := map[string]*http.Server{
-		"main":    server,
-		"admin":   adminServer,
-		"metrics": metricsServer,
+	mainServer := buildServer(ctx, env, healthState, probe, stats, logger)
+	secondaryServers := map[string]*http.Server{
+		"admin":   buildAdminServer(logger, healthState),
+		"metrics": buildMetricsServer(promStatReporter, protoStatReporter),
 	}
-
 	if env.EnableProfiling {
-		servers["profile"] = profiling.NewServer(profiling.NewHandler(logger, true))
+		secondaryServers["profile"] = profiling.NewServer(profiling.NewHandler(logger, true))
 	}
 
-	errCh := make(chan error, len(servers)+1)
-	listeners := make(map[string]net.Listener, len(servers))
-	for name, server := range servers {
-		l, err := net.Listen("tcp", server.Addr)
-		if err != nil {
-			logger.Fatalw("listen failed", zap.Error(err))
-		}
+	// Create the tcp listener for the main server before the unix listener to make sure
+	// the tcp serving is ready before unix serving.
+	mainListener, err := net.Listen("tcp", mainServer.Addr)
+	if err != nil {
+		logger.Fatalw("Failed to listen on %s for main server", mainServer.Addr, zap.Error(err))
+	}
 
-		listeners[name] = l
+	// Listen on a unix socket so that the exec probe can avoid having to go
+	// through the full tcp network stack.
+	unixListener, err := net.Listen("unix", unixSocketPath)
+	if err != nil {
+		logger.Fatalw("Failed to listen on unix socket", zap.Error(err))
 	}
 
-	for name, server := range servers {
+	errCh := make(chan error)
```

Comment:
I think this isn't safe because both of these could fail. Though if that happens we'll be doing a log.Fatal soon enough anyway so maybe it's fine

Reply:
It's actually 4 that can fail but I think it doesn't matter here as the goroutines hanging around won't damage anything as we'd exit immediately anyway or if we're already shutting down... well... we're already shutting down.